test/test_wallets.py

```diff
@@ -10,28 +10,139 @@
+                     wallet_structure, amount, wallet_file, password):
+    setup_import(mainnet=False)
+    wallet = make_wallets(1,[wallet_structure],
+                          fixed_seeds=[wallet_file],
+                          test_wallet=True, passwords=[password])[0]['wallet']
+    sync_count = 0
+    jm_single().bc_interface.wallet_synced = False
+    while not jm_single().bc_interface.wallet_synced:
+        jm_single().bc_interface.sync_wallet(wallet)
+        sync_count += 1
+        #avoid infinite loop
+        assert sync_count < 10
+        log.debug("Tried " + str(sync_count) + " times")
+
+    assert jm_single().bc_interface.wallet_synced
+    #do some transactions with the wallet, then close, then resync
+    for i in range(num_txs):
+        do_tx(wallet, amount)
+        log.debug("After doing a tx, index is now: " + str(wallet.index))
+        #simulate a spammer requesting a bunch of transactions. This
+        #mimics what happens in CoinJoinOrder.__init__()
+        for j in range(fake_count):
+            #Note that as in a real script run,
+            #the initial call to sync_wallet will
+            #have set wallet_synced to True, so these will
+            #trigger actual imports.
+            cj_addr = wallet.get_internal_addr(0)
+            change_addr = wallet.get_internal_addr(0)
+            wallet.update_cache_index()
+            log.debug("After doing a spam, index is now: " + str(wallet.index))
+
+    assert wallet.index[0][1] == num_txs+fake_count*2*num_txs
+
+    #Attempt re-sync, simulating a script restart.
+
+    jm_single().bc_interface.wallet_synced = False
+    sync_count = 0
+    #Probably should be fixed in main code:
+    #wallet.index_cache is only assigned in Wallet.__init__(),
+    #meaning a second sync in the same script, after some transactions,
+    #will not know about the latest index_cache value (see is_index_ahead_of_cache),
+    #whereas a real re-sync will involve reading the cache from disk.
+    #Hence, simulation of the fact that the cache index will
+    #be read from the file on restart:
+    wallet.index_cache = wallet.index
+
+    while not jm_single().bc_interface.wallet_synced:
+        #Wallet.__init__() resets index to zero.
+        wallet.index = []
+        for i in range(5):
+            wallet.index.append([0, 0])
+        #Wallet.__init__() also updates the cache index
+        #from file, but we can reuse from the above pre-loop setting,
+        #since nothing else in sync will overwrite the cache.
+
+        #for regtest add_watchonly_addresses does not exit(), so can
+        #just repeat as many times as possible. This might
+        #be usable for non-test code (i.e. no need to restart the
+        #script over and over again)?
+        sync_count += 1
+        log.debug("TRYING SYNC NUMBER: " + str(sync_count))
+        jm_single().bc_interface.sync_wallet(wallet)
+        #avoid infinite loop on failure.
+        assert sync_count < 10
```

Comment:
I'm inclined to believe that, as this comment suggests, we can do this for the main user level scripts. Just re-do sync several times until it works. Understood that it's not a trivial question. I do understand why the stop-restart approach is currently used: we can only pick up transactions with listtransactions from addresses that are already imported, so once we find out we have not imported enough, we need to go back to the beginning in some sense. But I believe doing it like this - resetting wallet.index to all zeros, picking up the cache from the file once only and keeping it (since it isn't modified during sync), and then just looping until wallet_synced is True (or some sanity iteration number for failure), might be viable.